libs/filer/workspace_fuse_client.go

```diff
@@ -0,0 +1,218 @@
+		entries[i] = newEntry
+
+		if _, ok := seenPaths[newEntry.Name()]; ok {
+			return nil, DuplicatePathError{
+				oi1:        seenPaths[newEntry.Name()],
+				oi2:        sysInfo,
+				commonName: path.Join(name, newEntry.Name()),
+			}
+		}
+		seenPaths[newEntry.Name()] = sysInfo
+
+	}
+
+	return entries, nil
+}
+
+// Note: There is loss of information when writing a ipynb file. A notebook written
+// with .Write(name = "foo.ipynb") will be written as "foo" in the workspace and
+// will have to be read as .Read(name = "foo.py") instead of "foo.ipynb
+//
+// Note: The import API returns opaque internal errors for namespace clashes
+// (e.g. a file and a notebook or a directory and a notebook). Thus users of this
+// method should be careful to avoid such clashes.
+func (w *workspaceFuseClient) Write(ctx context.Context, name string, reader io.Reader, mode ...WriteMode) error {
+	return w.workspaceFilesClient.Write(ctx, name, reader, mode...)
+}
+
+// Try to read the file as a regular file. If the file is not found, try to read it as a notebook.
+func (w *workspaceFuseClient) Read(ctx context.Context, name string) (io.ReadCloser, error) {
+	r, err := w.workspaceFilesClient.Read(ctx, name)
+
+	// If the file is not found, it might be a notebook.
+	if errors.As(err, &FileDoesNotExistError{}) {
+		stripName, _, ok := stripNotebookExtension(ctx, w.workspaceFilesClient, name)
+		if !ok {
+			// Not a valid notebook. Return the original error.
+			return nil, err
+		}
+		return w.workspaceFilesClient.Read(ctx, stripName)
+	}
+	return r, err
+}
+
+// Try to delete the file as a regular file. If the file is not found, try to delete it as a notebook.
+func (w *workspaceFuseClient) Delete(ctx context.Context, name string, mode ...DeleteMode) error {
+	err := w.workspaceFilesClient.Delete(ctx, name, mode...)
+
+	// If the file is not found, it might be a notebook.
+	if errors.As(err, &FileDoesNotExistError{}) {
+		stripName, _, ok := stripNotebookExtension(ctx, w.workspaceFilesClient, name)
+		if !ok {
+			// Not a valid notebook. Return the original error.
+			return err
+		}
+		return w.workspaceFilesClient.Delete(ctx, stripName, mode...)
+	}
+
+	return err
+}
+
+// Try to stat the file as a regular file. If the file is not found, try to stat it as a notebook.
+func (w *workspaceFuseClient) Stat(ctx context.Context, name string) (fs.FileInfo, error) {
+	info, err := w.workspaceFilesClient.Stat(ctx, name)
+
+	// If the file is not found in the local file system, it might be a notebook.
+	if errors.As(err, &FileDoesNotExistError{}) {
+		_, stat, ok := stripNotebookExtension(ctx, w.workspaceFilesClient, name)
+		if !ok {
+			return nil, err
+		}
+
+		// Since the file is a notebook, we should return the stat of the notebook,
+		// with the path modified to include the extension.
+		newStat := stat.(wsfsFileInfo)
+		newStat.oi.Path = newStat.oi.Path + notebook.GetExtensionByLanguage(&newStat.oi)
+		return newStat, nil
```

Comment:
Can you comment on the mtimes?

Reply:
We already parse and set the mtimes in the workspace_files_client, so it works out of the box and seems accurate based on my personal testing. 

The FUSE mount has mtimes missing for notebooks, but since we no longer rely on the FUSE mount we should be in the clear here.